crates/futures/src/task/multithread.rs

```diff
@@ -0,0 +1,201 @@
+use std::future::Future;
+use std::mem::ManuallyDrop;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::cell::RefCell;
+use std::sync::Arc;
+use std::sync::atomic::AtomicI32;
+use std::sync::atomic::Ordering::SeqCst;
+use std::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+use wasm_bindgen::JsCast;
+use wasm_bindgen::prelude::*;
+use crate::queue::Queue;
+
+
+#[wasm_bindgen]
+extern "C" {
+    type Atomics;
+
+    #[wasm_bindgen(static_method_of = Atomics, js_name = waitAsync)]
+    fn wait_async(buf: &JsValue, index: i32, value: i32) -> js_sys::Promise;
+
+    #[wasm_bindgen(static_method_of = Atomics, js_name = waitAsync, getter)]
+    fn get_wait_async() -> JsValue;
+}
+
+fn wait_async(ptr: &AtomicI32, current_value: i32) -> js_sys::Promise {
+    // If `Atomics.waitAsync` isn't defined (as it isn't defined anywhere today)
+    // then we use our fallback, otherwise we use the native function.
+    if Atomics::get_wait_async().is_undefined() {
+        crate::task::wait_async_polyfill::wait_async(ptr, current_value)
+
+    } else {
+        let mem = wasm_bindgen::memory().unchecked_into::<js_sys::WebAssembly::Memory>();
+        Atomics::wait_async(&mem.buffer(), ptr as *const AtomicI32 as i32 / 4, current_value)
+    }
+}
+
+fn notify(atomic: &AtomicI32, wakeup: u32) {
+    unsafe {
+        core::arch::wasm32::atomic_notify(
+            // TODO verify the safety of this
+            atomic as *const AtomicI32 as *mut i32,
+            wakeup, // Number of threads to notify
+        );
+    }
+}
+
+
+const SLEEPING: i32 = 0;
+const AWAKE: i32 = 1;
+
+struct AtomicWaker {
+    state: AtomicI32,
+}
+
+impl AtomicWaker {
+    fn new() -> Arc<Self> {
+        Arc::new(Self {
+            state: AtomicI32::new(AWAKE),
+        })
+    }
+
+    fn done(&self) {
```

Comment:
Could these one-liner methods be inlined into where they're used? I find it somewhat difficult to jump back and forth when most of them are only used once anyway

Reply:
I did it that way to encapsulate everything inside the `AtomicWaker`, so that way it's easier to understand, rather than having internal state updates strewn around the file.

But I don't have particularly strong opinions on it one way or the other.